Feature/data/product

hyun132 opened the pull request:
Firestore에서 데이터 조회 쿼리 시 발생하는 오류 수정

hyun132 committed:
fix[#19] : 의존성 수정

hyun132 committed:
feature [#19] di 사용 전 세팅, module설정

f-lab-dean reviewed (approved):
코드의 변화가 어떤 맥락인지 알기 어려운 것 같아요.
왜 테스트 코드와 관련된 의존성이 변경되었는지, MainActivity가 구현되었는지 이해하기 어렵습니다.
브랜치명과 커밋을 잘 활용하고, PR에는 하나의 공통된 작업으로 구성될 수 있도록 해주세요.
그리고 PR 작성 시 Description을 잘 작성해주세요.

hyun132 committed:
fix[#23] : Firestore에서 데이터 가져올 때 발생하는 오류 수정.

Product와 ProductPreview를 같은 /product collection에서 가져오고 있음
-> ProductPreview 목록 가져올 때 데이터를 ProductPreview타입으로 변환하려고 하느 부분에서 문제 발생
-> snapshot에서 데이터 받을 때는 일단 Product로 받아서 ProductPreview로 변환해주는 방식으로 수정

잘못된 collection명, 항목명 수정

hyun132 committed:
Revert "feature [#19] di 사용 전 세팅, module설정"

This reverts commit d74bb65f3d847ca2966f87f46cbe5a5caf9e9ac7.

hyun132 committed:
Revert "fix[#19] : 의존성 수정"

This reverts commit 03d324deefafe93f06241c5205e739c89c02e12c.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/ProductRepositoryImpl.kt:
```diff
@@ -38,25 +37,24 @@ class ProductRepositoryImpl @Inject constructor(
             .mapNotNull { snapshot ->  // null이면 무시
                 snapshot.firstOrNull()?.toObject(ProductDto::class.java)
-                    ?.let { productMapper.mapToProduct(it) }
+                    ?.let { productMapper.mapToProduct(it) }?: throw Exception("데이터가 없습니다.")
```
이럴땐 커스텀한 Exception을 정의하는 것도 좋습니다.
예를들면 `FireStoreNoDataException` 같이요.

hyun132 replied:
아 예외 처리하기 더 편할 것 같네요! 반영 해 보겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/ProductRepositoryImpl.kt:
```diff
@@ -38,25 +37,24 @@ class ProductRepositoryImpl @Inject constructor(
+            firestore.collection("product").snapshots()
                 .map { convertToProductList(it.toObjects(ProductPreviewDto::class.java)) }
+                .map {
+                    it.filter { it.name.contains(searchParam.keyword) }
```
람다의 it이 중복되면 어떤 것을 의미하는지 알기 어려울 수 있습니다.
IDE에서도 경고를 보여주고 있을 거에요.
스코프 마다 이름을 부여해서 명확하게 표현하는 것을 권장합니다.

hyun132 replied:
앗 그렇네요 수정하겠습니다!!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/ProductRepositoryImpl.kt:
```diff
@@ -38,25 +37,24 @@ class ProductRepositoryImpl @Inject constructor(
+                }
         } else {
-            firestore.collection("products").limit(20).snapshots()
+            firestore.collection("product").limit(20).snapshots()
```
product 라는 스트링이 반복된다면 상수로 선언하는 것을 권장합니다.
하드코딩보다는 상수가 항상 권장되고, 동일한 파라미터를 사용하는 곳을 상수의 변화만으로 모두 적용할 수 있어요.
실수로 놓친다던가, 오타라던가 등을 방지할 수 있습니다.

hyun132 replied:
넵 상수로 변경하겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/model/ProductDto.kt:
```diff
@@ -4,22 +4,43 @@ import com.squareup.moshi.JsonClass
+    val masterProductName: String? = "",
+    val dealProducts: List<Long>? = emptyList()
+) {
+    constructor() : this(
```
FireStore를 위해 디폴트 생성자를 만든 것이라면, 생성자를 추가하는 것 보다 파라미터에 디폴트 값을 주는건 어떠세요?
이미 대부분의 파라미터에 디폴트 값이 있기 때문에 생성자를 추가하는 것이 큰 의미가 없어보입니다.

hyun132 replied:
그렇네요,, 수정하겠습니다!

hyun132 committed:
fix[#23] : [피드백 반영] collection path 상수로 선언, 중첩되는 스코프의 파라미터 이름 설정, FireStoreNoDataException 생성

hyun132 committed:
fix[#23] : [피드백 반영] 디폴트 값 추가 및 불필요한 생성자 제거

f-lab-dean reviewed (approved):
LGTM!
